Update the two qubit test

nwittler opened the pull request:
## What
Cleanup the integration test for the two qubit device.

## Why
It was a mess.

## How
Moved setup code to proper pytest fixtures, including parametric testing for all search algorithms (currently limited to verified ones)

## Remarks
This should be a simpler, more reliable test that includes fewer moving parts. The algorithm library should be cleaned and included in the parametric testing.

## Checklist
Please include and complete the following checklist. Your Pull Request is (in most cases) not ready for review until the following have been completed. You can create a draft PR while you are still completing the checklist. Check the [Contribution Guidelines](https://github.com/q-optimize/c3/blob/dev/CONTRIBUTING.md) for more details. You can mark an item as complete with the `- [x]` prefix

- [x] Tests - Added unit tests for new code, regression tests for bugs and updated the integration tests if required
- [x] Formatting & Linting - `black` and `flake8` have been used to ensure styling guidelines are met
- [x] Type Annotations - All new code has been type annotated in the function signatures using type hints
- [ ] Docstrings - Docstrings have been provided for functions in the `numpydoc` style
- [x] Documentation - The tutorial style documentation has been updated to explain changes & new features
- [x] Notebooks - Example notebooks have been updated to incorporate changes and new features
- [ ] Changelog - A short note on this PR has been added to the Upcoming Release section

nwittler committed:
Removed deco from unverified algorithms

nwittler committed:
Simplified and refactored test and data

nwittler committed:
Put needed decos for cma variants back.

nwittler committed:
Docstrings.

lazyoracle commented:
> This should be a simpler, more reliable test

What do you have in mind when you say more reliable? Do you mean this test fails when something _isn't_ wrong or that the pass/fail of the test is not reproducible? If that is the case, then we definitely should look deeper into it. If you mean it is difficult to debug this test, because it seems like this is the only test that fails while other tests pass, that in most cases is indicative of the fact that we do not have the necessary _unit_ tests covering the buggy section of the code and not of any inherent issue with this integration test. 

> that includes fewer moving parts

I am not entirely sure if we want this test to have fewer moving parts. As an integration test, this is expected to interface as many moving parts as possible to ensure the whole big machinery works as expected in a typical workflow. The example notebooks also provide a similar level of integration testing but it is usually not part of the development workflow.

lazyoracle committed:
maxiters for tf_sgd, maxiter otherwise

In general, optimization algorithms should not be expected to have the
same set of options, more so when they come different libraries. So the
expected usage is to provide the options distinctly for every algorithm.

lazyoracle committed:
add decorator to sweep, reqd for sensitivity

lazyoracle commented:
a21ed101af4aeb668ea5c2435b26d5f7f2f8e84a - I am not sure if this is the right approach. The decorator is a perfectly valid way to create the library of available algorithms. Checks and balances to ensure that people don't use algorithms that are not tested could be put in place through `Warning`s and `Error`s. 

I would rather use the `pytest` fixture to decide which algorithms need/are ready to be tested, and make it explicit in the tests by defining the algorithm list inside the fixture, rather than excluding algorithms from being added to the library with the decorator. This is because the algorithms library contains algorithms that are not all meant for Optimal Control workflows (eg `sweep`, `adaptive_scan` etc). This is the reason behind the failing `test_sensitivity` and `test/test_optim_init.py::test_main_sens`.

lazyoracle committed:
explicitly mention algos to be tested

lazyoracle committed:
increase tolerance

lazyoracle commented on test/test_two_qubits.py:
```diff
@@ -2,445 +2,71 @@
-
 
-def test_signals() -> None:
+def test_signals(get_two_qubit_chip) -> None:
```
add relevant `pytest` markers (`slow`, `integration`, `unit` etc) to this test

lazyoracle commented on test/test_two_qubits.py:
```diff
@@ -2,445 +2,71 @@
 
 
-def test_hamiltonians() -> None:
+def test_hamiltonians(get_two_qubit_chip) -> None:
```
`pytest` markers

lazyoracle commented on test/test_two_qubits.py:
```diff
@@ -2,445 +2,71 @@
 
 
-def test_propagation() -> None:
+def test_propagation(get_two_qubit_chip) -> None:
```
`pytest` markers

lazyoracle commented on test/test_two_qubits.py:
```diff
@@ -2,445 +2,71 @@
 
 
-def test_init_point() -> None:
+def test_init_point(get_OC_optimizer) -> None:
```
`pytest` markers

lazyoracle commented on test/test_two_qubits.py:
```diff
@@ -2,445 +2,71 @@
 
 
-def test_rk4() -> None:
+def test_rk4(get_two_qubit_chip) -> None:
```
`pytest` markers, and maybe include an assert statement instead of just relying on this raising an error if RK4 is not working?

nwittler committed:
Re-decorated.

nwittler committed:
pytest decos

lazyoracle reviewed (approved):
LGTM
